ethereum/contracts/relayer/coreRelayer/CoreRelayerGovernance.sol

```diff
@@ -0,0 +1,208 @@
+// SPDX-License-Identifier: Apache 2
+pragma solidity ^0.8.19;
+
+import "@openzeppelin/contracts/proxy/ERC1967/ERC1967Upgrade.sol";
+
+import {IWormhole} from "../../interfaces/IWormhole.sol";
+import {InvalidPayloadLength} from "../../interfaces/relayer/IWormholeRelayer.sol";
+import {fromWormholeFormat} from "./Utils.sol";
+import {BytesParsing} from "./BytesParsing.sol";
+import {
+  getGovernanceState,
+  getRegisteredCoreRelayersState,
+  getDefaultRelayProviderState
+} from "./CoreRelayerStorage.sol";
+import {CoreRelayerBase} from "./CoreRelayerBase.sol";
+
+error GovernanceActionAlreadyConsumed(bytes32 hash);
+error InvalidGovernanceVM(string reason);
+error InvalidGovernanceChainId(uint16 parsed, uint16 expected);
+error InvalidGovernanceContract(bytes32 parsed, bytes32 expected);
+
+error InvalidPayloadChainId(uint16 parsed, uint16 expected);
+error InvalidPayloadAction(uint8 parsed, uint8 expected);
+error InvalidPayloadModule(bytes32 parsed, bytes32 expected);
+error InvalidFork();
+error ConractUpgradeFailed(bytes failure);
+
+abstract contract CoreRelayerGovernance is CoreRelayerBase, ERC1967Upgrade {
+//TODO AMO: the bytes32 payloads here are suspect - for EVM we only need 20 bytes
+//           for other chains, might it not take more than just a single address?
+//           Wouldn't it be safer to just allow arbitrary bytes at the end of the payload?
+
+//governance payload structs (packed)
+//   struct ContractUpgrade {
+//     bytes32 module;
+//     uint8 action;
+//     uint16 chainId;
+//     bytes32 newImplementation; //TODO AMO: why is this not an address?
+//   }
+//
+//   struct RegisterChain {
+//     bytes32 module;
+//     uint8 action;
+//     uint16 chainId; //TODO Why is this on this object? //TODO AMO: why indeed?
+//     uint16 emitterChain;
+//     bytes32 emitterAddress;
+//   }
+//
+//   //This could potentially be combined with ContractUpgrade
+//   struct UpdateDefaultProvider {
+//     bytes32 module;
+//     uint8 action;
+//     uint16 chainId;
+//     bytes32 newProvider; //TODO AMO: why is this not an address?
+//   }
+
+  //TODO AMO: What's this module business about? (and also the action business for that matter)
```

Comment:
Module is 'CoreRelayer' keccak256-ed. (perhaps this should be written as a comment, hmm)

Reply:
Btw, I'd be in favor of just setting the `module` constant to `bytes32(CoreRelayer)` :P

Reply:
AFAIK, module adds domain information for the governance VAA and action, the same but as a way to choose a particular governance action for that module. It's an entirely custom dispatch for contract interaction. Another option could be to use the Ethereum ABI, but we would be imposing that on other VMs and that may end up being a bit impractical. I don't know which parts of this are flexible and which are not in regards to the guardian codebase.

Reply:
It's actually not the same as `bytes32("CoreRelayer")` because the string version is left-aligned while the hex number version is right-aligned.

Reply:
I know. I just don't see any purpose in choosing a particular alignment, it just obfuscates what it is :pensive:

Reply:
True. the problem is the guardian code for upgrade/register chain has the module left-padded so this would need to be the case in order to reuse that :(